### DISCUSSION TITLE
"Logical" Error Correction

### DISCUSSION THREAD

**Discussion by dev8-de** [jomjol/AI-on-the-edge-device#1170] [Q&A]
Hi,

is there a way to implement a "logical" error correction based on single digits due to faulty recognition?

For Example:

![gas](https://user-images.githubusercontent.com/11597226/195707370-134c0bc9-b4b3-4583-b844-134d777c3d9f.jpg)

As you can see, the Raw Value is wrong, the first digit befor the decimal point is an 8, but a 3 is readout, so I got a negative error.

My Idea is now, that there will be a logical error correction based on the digi. In this example:

3 will be changed with 8 and vice versa and then will be checked if the new raw value is within rate rate defined in the config (eg. 0.5):

35803.830 - 35808.587 is negative.
85803.830 - 35808.587 > 0.5
35308.830 - 35808.587 is negative
35803.830 - 35803.587 < 0.5 - new value

If there is not plan to implement it, if there is an interest? My C/CPP Skills are limited, but I think I'll can do it. ;)

Regards,
  easy.

**Comment by ghost** [comment]
Look at recognition. I think, the 8 will not recognized correctly from the models. You can try to make the ROIs a bit smaller like the ROI after the comma.
The reflection can be a reason for the false prediction. You could tape the reflection of the led with a dark tape. 

Is nearly the same gasometer and I have no false predictions.
<img width="443" alt="image" src="https://user-images.githubusercontent.com/412645/195710314-b4b8016a-b264-4945-b749-345bb3a207f0.png">

**Comment by dev8-de** [comment]
Well, I already have black foam over the LED, without it (or like white paper or somehing like that) the last digit befor the comma wouldn't be readable at all.

My problem is, that the reading changes like desribed (mqtt):

{'value': 35810.907, 'raw': '35810.907', 'error': 'no error', 'rate': 0.003299, 'timestamp': '2022-10-14T08:20:30'}
{'value': 35810.947, 'raw': '35810.947', 'error': 'no error', 'rate': 0.008247, 'timestamp': '2022-10-14T08:25:21'}
{'value': 35810.947, 'raw': '35810.947', 'error': 'no error', 'rate': 0.0, 'timestamp': '2022-10-14T08:30:12'}
{'value': 35810.962, 'raw': '35810.962', 'error': 'no error', 'rate': 0.003093, 'timestamp': '2022-10-14T08:35:03'}
{'value': '', 'raw': '35871.002', 'error': 'Rate too high - Read: 35871.002 - Pre: 35810.962', 'rate': '', 'timestamp': '2022-10-14T08:35:03'}
{'value': '', 'raw': '35871.002', 'error': 'Rate too high - Read: 35871.002 - Pre: 35810.962', 'rate': '', 'timestamp': '2022-10-14T08:35:03'}
{'value': '', 'raw': '35871.016', 'error': 'Rate too high - Read: 35871.016 - Pre: 35810.962', 'rate': '', 'timestamp': '2022-10-14T08:35:03'}

As you can see, the second last digit befor the comma doesn't changed, but as soon as the last digit flips from 0 to 1, the recognation for the second last goes wrong, without changing anything. And this happens alot.

Also in the setup for the ROI are 2 rectangles per digit. What is the important one? The outer or the inner rectangle?

Greetings.

**Comment by dev8-de** [comment]
Addition: I have even a flip if the last digit befor the comma doesn't change:

{'value': '', 'raw': '35871.057', 'error': 'Rate too high - Read: 35871.057 - Pre: 35810.962', 'rate': '', 'timestamp': '2022-10-14T08:35:03'}
{'value': '', 'raw': '35871.069', 'error': 'Rate too high - Read: 35871.069 - Pre: 35810.962', 'rate': '', 'timestamp': '2022-10-14T08:35:03'}
{'value': 35811.11, 'raw': '35811.110', 'error': 'no error', 'rate': 0.004359, 'timestamp': '2022-10-14T09:09:00'}
{'value': '', 'raw': '35871.110', 'error': 'Rate too high - Read: 35871.110 - Pre: 35811.110', 'rate': '', 'timestamp': '2022-10-14T09:09:00'}
{'value': '', 'raw': '35871.121', 'error': 'Rate too high - Read: 35871.121 - Pre: 35811.110', 'rate': '', 'timestamp': '2022-10-14T09:09:00'}

and so on.. So you can see, sometimes the recognition is correct sometime not..

**Reply by ghost** [reply]
Look at the ROIs of my screenshot. A good ROI config is essential. A screenshot of the recognition page could help and wich model you use.
